src/main/java/hudson/plugins/jobConfigHistory/FileHistoryDao.java

```diff
@@ -712,9 +709,46 @@ private File getJobDirectoryIncludingFolder(final String folderName) {
 		final String realFolderName = folderName.isEmpty()
 				? folderName
 				: folderName + "/jobs";
+		getFolderFiles();
 		return new File(getJobHistoryRootDir(), realFolderName);
 	}
 
+	private File[] getJobsIncludingThoseInFolders(final FileFilter fileFilter) {
+		final List<File> folderFiles = getFolderFiles();
+		File[] standardJobs = getJobHistoryRootDir().listFiles(fileFilter);
+
+		for (int i = 0; i < folderFiles.size(); ++i) {
+			if (!fileFilter.accept(folderFiles.get(i))) {
+				folderFiles.remove(folderFiles.get(i));
+			}
+		}
+		if (standardJobs != null) {
+			folderFiles.addAll(Arrays.asList(standardJobs));
+		}
+		return folderFiles.toArray(new File[folderFiles.size()]);
+	}
+
+	private List<File> getFolderFiles() {
+		//no full tree search, just one layer
+		List<File> folderNames = new LinkedList<File>();
+		File jobHistoryRootDir = getJobHistoryRootDir();
+		File[] jobHistoryRootDirFiles = jobHistoryRootDir.listFiles();
+		if (jobHistoryRootDirFiles != null) {

```

Comment:
Too many nested ifs/loops. Instead of
```
for(...){
  if(...){
    for(...){
      ...
    }
  }
}
```
you can do sth like this: 
```
for(...){
  if(!...){
   continue;
  }
  for(...){
   ...
  }
}
```

Reply:
Done.